### ISSUE TITLE
Customer and Subscription POST now return a `data` array in the JSON, which can't be de-serializaed to proper objects

### ISSUE THREAD

**Issue by nicolaiarocci** [stripe/stripe-mock#170]
We updated to latest stripe-mock/stripe-dontnet releases (0.60/27.10). Now both Customer and Subscription POST responses return an array of objects (the `data` field), whereas they used to return the newly added object. This appears not to be coherent with Stripe API, which still returns the object, and not an array. A quick check at the OpenAPI specification seems to confirm that a single object is still expected.

This is the `StripeResponse.Content` content following a `Stripe.CustomerService.Create()` call. The `Stripe.Customer` instance returned by this call is a non-initialized object (e.g. `Customer.Id` is null, like all other fields.)

```
"{\"data\":[{\"account_balance\":0,\"address\":null,\"balance\":0,\"created\":1234567890,\"currency\":\"usd\",\"default_source\":null,\"delinquent\":false,\"description\":null,\"discount\":null,\"email\":null,\"id\":\"cus_FJaYLWmiE8rxFC\",\"invoice_prefix\":\"180BC08\",\"invoice_settings\":{\"custom_fields\":null,\"default_payment_method\":null,\"footer\":null},\"livemode\":false,\"metadata\":{},\"name\":null,\"object\":\"customer\",\"phone\":null,\"preferred_locales\":[],\"shipping\":null,\"sources\":{\"data\":[{\"created\":1234567890,\"customer\":null,\"fingerprint\":\"OUv5A7ge4czHxnfR\",\"id\":\"aliacc_1EphH5IUzFSXAEcAfYCgLj9h\",\"livemode\":false,\"metadata\":{},\"object\":\"alipay_account\",\"payment_amount\":1000,\"payment_currency\":\"usd\",\"reusable\":false,\"used\":false,\"username\":\"test@example.com\"}],\"has_more\":false,\"object\":\"list\",\"url\":\"/v1/customers/cus_FJaYLWmiE8rxFC/sources\"},\"subscriptions\":{\"data\":[{\"application_fee_percent\":null,\"billing\":\"charge_automatically\",\"billing_cycle_anchor\":1234567890,\"billing_thresholds\":null,\"cancel_at_period_end\":false,\"canceled_at\":1234567890,\"collection_method\":\"charge_automatically\",\"created\":1234567890,\"current_period_end\":1234567890,\"current_period_start\":1234567890,\"customer\":\"cus_FJaYLWmiE8rxFC\",\"days_until_due\":null,\"default_payment_method\":null,\"default_source\":null,\"default_tax_rates\":[],\"discount\":null,\"ended_at\":1234567890,\"id\":\"sub_FKLz5hl0zyaPGd\",\"items\":{\"data\":[{\"billing_thresholds\":null,\"created\":1561579163,\"id\":\"si_FKLzxZRXl8f8qr\",\"metadata\":{},\"object\":\"subscription_item\",\"plan\":{\"active\":true,\"aggregate_usage\":null,\"amount\":2000,\"billing_scheme\":\"per_unit\",\"created\":1561402737,\"currency\":\"usd\",\"id\":\"gold\",\"interval\":\"month\",\"interval_count\":1,\"livemode\":false,\"metadata\":{},\"nickname\":null,\"object\":\"plan\",\"product\":\"prod_FJaYU37yVcOAWH\",\"tiers\":null,\"tiers_mode\":null,\"transform_usage\":null,\"trial_period_days\":null,\"usage_type\":\"licensed\"},\"quantity\":1,\"subscription\":\"sub_FKLzukkMx7NSmO\"}],\"has_more\":false,\"object\":\"list\",\"url\":\"/v1/subscription_items?subscription=sub_FKLz5hl0zyaPGd\"},\"latest_invoice\":null,\"livemode\":false,\"metadata\":{},\"object\":\"subscription\",\"plan\":{\"active\":true,\"aggregate_usage\":null,\"amount\":2000,\"billing_scheme\":\"per_unit\",\"created\":1561402737,\"currency\":\"usd\",\"id\":\"gold\",\"interval\":\"month\",\"interval_count\":1,\"livemode\":false,\"metadata\":{},\"nickname\":null,\"object\":\"plan\",\"product\":\"prod_FJaYU37yVcOAWH\",\"tiers\":null,\"tiers_mode\":null,\"transform_usage\":null,\"trial_period_days\":null,\"usage_type\":\"licensed\"},\"quantity\":1,\"start\":1234567890,\"start_date\":1234567890,\"status\":\"active\",\"tax_percent\":null,\"trial_end\":1234567890,\"trial_start\":1234567890}],\"has_more\":false,\"object\":\"list\",\"url\":\"/v1/customers/cus_FJaYLWmiE8rxFC/subscriptions\"},\"tax_exempt\":\"none\",\"tax_ids\":{\"data\":[{\"country\":\"DE\",\"created\":1234567890,\"customer\":\"cus_FJaYLWmiE8rxFC\",\"id\":\"txi_123456789\",\"livemode\":false,\"object\":\"tax_id\",\"type\":\"eu_vat\",\"value\":\"DE123456789\",\"verification\":{\"status\":\"pending\",\"verified_address\":null,\"verified_name\":null}}],\"has_more\":false,\"object\":\"list\",\"url\":\"/v1/customers/cus_FJaYLWmiE8rxFC/tax_ids\"},\"tax_info\":null,\"tax_info_verification\":null}],\"has_more\":false,\"object\":\"list\",\"url\":\"/v1/customers\"}"
```

Not sure if this is the intended behaviour, and then I am obviously missing something in my upgrade path.

**Comment by ob-stripe** [comment]
Hi @nicolaiarocci, can you check your requests and share the exact path you're using?

I just tested it on my laptop and it seems to work fine:
```
$ stripe-mock -version
stripe-mock 0.60.0

$ curl http://localhost:12111/v1/customers -H "Authorization: Bearer sk_test_123" -X POST -d email=foo@example.com
{
  ...
  "id": "cus_Czk27xmfTfCoyfK",
  ...
  "object": "customer",
  ...
}

$ curl http://localhost:12111/v1/subscriptions -X POST -H "Authorization: Bearer sk_test_123" -d customer=cus_123
{
  ...
  "id": "sub_Czk4BHM0z0YHrGj",
  ...
  "object": "subscription",
  ...
}
```

My guess is that you're actually sending GET requests, not POST, which causes stripe-mock to (correctly) return a list object.

**Comment by nicolaiarocci** [comment]
Hi @ob-stripe and thanks for looking into it. As mentioned, I am using `Stripe.CustomerService.Creare()` to perform the request. Same code executed against the Stripe test API returns a properly initialized `Customer` object. Will double check when I get back to my machine, and report back.

**Comment by nicolaiarocci** [comment]
Alright, I think I found the culprit. Long story short:

- if we set `StripeClient.ApiBase = "http://localhost:12111/"`, then a `Stripe.CustomerService.Create()` call will return a list;
- if we set `StripeClient.ApiBase = "http://localhost:12111"`, the same call will return an object.

It wasn't immediately apparent because we what we are actually doing is:

```cs
            var stripeClient = new Stripe.StripeClient(
                StripeConfiguration.ApiKey,
                apiBase: StripeConfiguration.BaseAddress.ToString()
                filesBase: StripeConfiguration.BaseAddress.ToString()
            );
            Stripe.StripeConfiguration.StripeClient = stripeClient;
```
Where `StripeConfiguration.BaseAddress` is a URI, and `ToString()` returns the trailing slash.

The same code above will always return a single object when `BaseAddress` is set to the remote Stripe API URI.

**Comment by ob-stripe** [comment]
Nice find!

We could fix in the Stripe.net client library by stripping the trailing `/`, but we should probably also fix this in stripe-mock (or at least return a better error to make it clear that `//v1/foo` is not a valid path).

@brandur-stripe Could you take a look at what's causing stripe-mock to return a list object in this case?

**Comment by nicolaiarocci** [comment]
Just out of curiosity I tried:

`curl -i http://localhost:12111//v1/customers -H "Authorization: Bearer sk_test_123"`

It returns a 301 which points to `/v1/customers`

**Comment by brandur-stripe** [comment]
Thanks for the diagnosis @nicolaiarocci and assist @ob-stripe! Sorry about the delayed response here.

Looking at this one a little more closely, it turns out that the redirect behavior you're seeing is built right into [Go's HTTP package](https://stackoverflow.com/questions/47475802/golang-301-moved-permanently-if-request-path-contains-additional-slash) and activates when double slashes are encountered in the URL. So the 301 is sent before the request even makes it into stripe-mock's HTTP handler.

Agreed though that we should be maximally forgiving when it comes to this sort of thing. It looks like there's a way to work around the default behavior, so going to try to put together a patch for that.

**Comment by brandur-stripe** [comment]
Released a fix in 0.62.0.
